docs/config.md

# Configuration Guide

The Multi-Source Data Store is configured using a YAML file. This configuration 
defines key elements such as datasets, data stores, grid mappings, and general 
parameters, which are essential for setting up a workflow to generate data cubes. 
This guide will walk you through the schema for each section of the configuration 
file and show you how to create a valid configuration.

To view the schema of the configuration, you can run the following code:

```python
from xcube_multistore.multistore import MultiSourceDataStore

MultiSourceDataStore.get_config_schema()
```

In the notebook [setup_config.ipynb](https://github.com/xcube-dev/xcube-multistore/blob/main/examples/setup_config.ipynb), you can find
an example demonstrating how to use auxiliary functions to set up a configuration.  
  
Example configurations can be explored in the [examples](https://github.com/xcube-dev/xcube-multistore/blob/main/examples) folder.

## Configuration Generator GUI

> **Note:** This feature is under active development, and only a minimal working
> example is currently available.

The **Configuration Generator GUI** shall provide an interactive interface for creating and
editing the configuration YAML, making the setup process more intuitive and less
error-prone.

**Key features (in development):**

- Display of all available fields for each configuration section
- Dynamic fetching and updating of valid parameters and inputs
- Dropdown menus that show only supported options
- Autofill assistance for large option sets (e.g., thousands of data IDs)
- Built-in configuration validator/checker
- Geolocation visualization to help define bounding boxes

To launch the GUI, run the following command from the package root:

```bash
panel serve xcube_multistore/gui/app.py --dev
```


## Entire configuration schema

The entire configuration schema consists of the following sections: 

**Properties:**

* **datasets** (Required): list of [dataset objects](#dataset-object) objects.
* **preload_datasets** (Optional): list of [preload dataset objects](#preload-dataset-object).
* **data_stores** (Required): list of [store objects](#store-object). A single data store with identifier `storage` is required for writing the final data cubes.
* **grid_mappings** (Optional): list of [grid mapping objects](#grid-mapping-object).
* **general** (Optional): following the [general object](#general-object) mapping.


## Object parameters

An object in this configuration schema can refer to a dataset, data store, or grid 
mapping. 

### dataset object

Two different dataset types are available:

#### single dataset object
This configures a dataset object from a single data source.  

**Properties**:

* **[identifier](#identifier)**: This identifier defines the name of the final data cube.
* **[store](#store)**
* **[grid_mapping](#grid_mapping)**
* **[data_id](#data_id)**
* **[open_params](#open_params)**
* **[format_id](#format_id)**
* **[chunksize](#chunksize)**
* **[custom_processing](#custom_processing)**
* **[spatial_resample_params](#spatial_resample_params)**
* **[temporal_resample_params](#temporal_resample_params)**


#### multi dataset object
This configures a datasets from multiple data sources. Each [variable object](#variable-object) is 
opened individually, and once all datasets are loaded, they are merged using 
`xarray.merge()`.

**Properties**:

* **[identifier](#identifier)**: This identifier defines the name of the final data cube.
* **[grid_mapping](#grid_mapping)**
* **[format_id](#format_id)**
* **[chunksize](#chunksize)**
* **[xr_merge_params](#xr_merge_params)**
* **variables**: List of [variable object](#variable-object) objects.

#### variable object
This configures a variable object, which is stored as a data variable within a single 
`xarray.Dataset` object.

**Properties**:

* **[identifier](#identifier)**: This identifier defines the variable name within the `xarray.Dataset`.
* **[store](#store)**
* **[data_id](#data_id)**
* **[open_params](#open_params)**
* **[custom_processing](#custom_processing)**
* **[spatial_resample_params](#spatial_resample_params)**
* **[temporal_resample_params](#temporal_resample_params)**

### store object
This configures a store object representing one xcube data store instance.

**Properties:**

* **[identifier](#identifier)**
* **[store_id](#store_id)**
* **[store_params](#store_params)**


### grid mapping object
This configures a grid mapping object.

**Properties:**

* **[identifier](#identifier)**
* **[bbox](#bbox)**
* **[crs](#crs)**
* **[spatial_res](#spatial_res)**
* **[tile_size](#tile_size)**

### preload dataset object
This object configures the preloading of datasets that require it. Note that some 
datasets are not accessible instantly via lazy loading, due to storage in compressed 
formats or due to lengthy queuing processes during data requests. These dataset can 
be preloaded once, which are then available to the user throughout the lifespan of
a project.

Currently, the preload API is available for the following data stores:

* [clms](https://github.com/xcube-dev/xcube-clms)
* [zenodo](https://github.com/xcube-dev/xcube-zenodo)

**Properties**:

* **[store](#store)**: The data store where the dataset is located.
* **data_ids**: A list of [data_id](#data_id) values for the datasets to preload.
* **preload_params**: Parameters specific to the respective data store for preloading.


### general object
This object configures general parameters that control various aspects of the cube 
generation process.

**Properties:**

* **[visualize](#visualize)**
* **[force_preload](#force_preload)**
* **[gdal_http_params](#gdal_http_params)**


## Single field parameters

### identifier
Identifier for any object like dataset, grid mapping, store.

### store
Store identifier used to open the dataset.

### grid_mapping
Identifier that assigns a grid mapping to the final dataset for reprojection.

### data_id
Unique identifier for the dataset's data source within the assigned data store.

### open_params
Open data parameters related to the data store and data_id.

### xr_merge_params
Parameters of `xarry.merge` needed if harmonization of multiple datasets into one
datacube is required. For further information view the [xarray.merge documentation](https://docs.xarray.dev/en/stable/generated/xarray.merge.html).
**Defaults**: `(compat="no_conflicts", join="exact", fill_value="<NA>", combine_attrs="drop_conflicts")`.

### format_id
Desired format of the saved datacube.  

**Default:** `zarr`  
**Allowed values:** `netcdf`, `zarr`, [`levels`](https://xcube.readthedocs.io/en/latest/mldatasets.html#the-xcube-levels-format)

### chunksize
Optional mapping from dimension to chunk size.

### spatial_resample_params
This section enables user to define the parameters for spatial resampling of 
the dataset. For a full list of supported parameters, refer to the
[`xcube_resampling.resample_in_space` Python API documentation](https://xcube-dev.github.io/xcube-resampling/api/#xcube_resampling.resample_in_space).


_NOTE_: The first 3 arguments `(source_ds, target_gm, source_gm)` in the
`xcube_resampling.resample_in_space` are not required and will be inferred
automatically from the config file's dataset and gridmapping section.

### temporal_resample_params
This section enables user to define the parameters for temporal resampling of
the dataset. For a full list of supported parameters, refer to the
[`xcube_resampling.resample_in_time` Python API documentation](https://xcube-dev.github.io/xcube-resampling/api/#xcube_resampling.resample_in_time).

_NOTE_: The argument `frequency` is required; the remaining keyword arguments are
optional.

### custom_processing
This section enables users to define a Python function that is executed after opening
the data. It allows for custom dataset manipulation, making it particularly useful
for handling unstructured datasets. The function must accept an `xarray.Dataset` as
input and return a modified `xarray.Dataset` as output.

**Properties**:  

* **module_path**: Path to the Python module relative to the config file.  
* **function_name**: Name of the function within the specified module. The function must
  take an `xarray.Dataset` as input and return a transformed `xarray.Dataset`.  

### store_id
The **store identifier** is used within the xcube ecosystem to specify different data 
stores. For more information about the available data stores, refer to the 
[xcube data store documentation](https://xcube.readthedocs.io/en/latest/dataaccess.html#data-store-framework),
or check the README of each specific data store plugin on GitHub, where links are given 
below.

**Allowed Values**:  

* [cds](https://github.com/xcube-dev/xcube-cds)  
* [clms](https://github.com/xcube-dev/xcube-clms)  
* [cmems](https://github.com/xcube-dev/xcube-cmems)  
* [esa-cci, esa-cci-kc, esa-cci-zarr](https://github.com/esa-cci/xcube-cci)
* [file, https, memory, s3](https://xcube.readthedocs.io/en/latest/dataaccess.html#filesystem-based-data-stores)   
* [sentinelhub, sentinelhub-cdse](https://github.com/xcube-dev/xcube-sh)  
* [smos](https://github.com/xcube-dev/xcube-smos)  
* [stac, stac-cdse, stac-cdse-ardc, stac-pc, stac-pc-ardc, stac-xcube](https://github.com/xcube-dev/xcube-stac)
* [zenodo](https://github.com/xcube-dev/xcube-zenodo)  

---

### store_params
The **store-specific parameters** define the configuration for a particular data store. 
These parameters can be retrieved using `get_data_store_params_schema(store_id)`. 
For more details, refer to the [xcube data store documentation](https://xcube.readthedocs.io/en/latest/dataaccess.html#data-store-framework), 
or consult the README of each specific data store plugin on GitHub, which is linked in 
the [store_id section](#store_id).

### spatial_res
Spatial Resolution of the final grid mapping.

### crs
The Coordinate Reference System (CRS) defines the coordinate reference system 
of the final grid mapping. Note that the CRS should be provided as a string, 
e.g., "EPSG:4326".

### bbox
The bounding box of the final grid mapping, expressed in the units of the CRS, is 
provided as [west, south, east, north].

### tile_size
The spatial chunk size in the grid mapping. If a single integer is provided, a square 
tile size is assumed. Otherwise, specify a list in the format [x, y] or [lon, lat].

**Default**: [1024, 1024]

### visualize
Switch between visualization in table, if True, and logging, if False.  

**Default:** `True`

### force_preload
If True, all data IDS given in section `preload_datasets` will be preloaded. If False, 
only non-preloaded datasets will be preloaded.  

**Default:** `False`

### gdal_http_params
GDAL http environment variables which are used when opening a tif file with 
`rioxarray.open_rasterio`, which uses GDAL driver under the hood, from a remote source.  

**Properties**:  

* **gdal_http_max_retry**: Maximal number of retries of an HTTP request in GDAL. Defaults set to `10`.  
* **gdal_http_retry_delay**: Delay in seconds between retries of an HTTP request in GDAL. Defaults set to `5`.   






CHANGES.md
## Changes in 0.3.4 (under development)

- Improved support for parallel multi-source datastore operations. Sentinel-2 and
  Sentinel-3 requests now use a randomly generated UUID for temporary file storage,
  preventing conflicts between concurrent requests. 
- Added a retry mechanism for the data stores `stac-cdse-ardc` and `stac-pc-ardc`
  to handle temporary STAC API and data access errors.


## Changes in 0.3.3

- Improved documentation:
  - Added a new "How-to" guide.
  - Expanded and improved the Data Availability page.
  - Update all examples in `examples` and `docs/examples`
- Reduced dependencies by removing the direct dependency on xcube data store plugins.


## Changes in 0.3.2

- Improved logging output when `visualize` is disabled in the configuration.
- Updated dependency versions for the xcube data store plugins.


## Changes in 0.3.1

- Use `xcube_resampling.gridmapping.GridMapping.regular_from_bbox` to create a 
  regular grid mapping instance from configuration.
- Added support for ERA5 time series extraction via the xcube-cds data store plugin.
- Improved documentation by adding a new Data Availability page.
- Fixed an issue by unifying chunking of concatenated datasets prior to writing.


## Changes in 0.3.0

* **Unified spatial resampling**: Switched to the new
  `xcube_resampling.resample_in_space` library, replacing
  `xcube.core.resampling.resample_in_space`.
* **Spatial resample parameters**: Added support for `spatial_resample_params`
  and updated the configuration visualization to display these parameters.
* **Temporal resampling**: Introduced `xcube_resampling.resample_in_time` with
  `temporal_resample_params` added to the configuration to cover temporal resampling.
* **Controlled cube generation**: `xcube_multistore.MultiSourceDataStore` is now only
  initialized in the constructor without automatically starting cube generation.
  A new `generate()` method is provided to explicitly start cube creation.
* **Geolocation visualization**: Added `display_geolocations()` to visualize
  all requested geolocations of the datacubes to be generated.
* Fixed visualization of preload handle
* Experimental configuration setup GUI added using panel.
* Implemented a new access handler that supports large dataset requests by
  automatically splitting and retrieving data from the `"cds"`, `"stac-cdse-ardc"`,
  and `"stac-pc-ardc"` data stores.


## Changes in 0.2.0

* Added the method `display_config`, which provides an overview of the cube generation 
  plan retrieved from the configuration file.  
* Added auxiliary functions to support setting up configurations for cube generation. 
  Examples of all functionalities are provided in the notebook [setup_config.ipynb](examples/setup_config.ipynb).


## Changes in 0.1.0

* The initial version of the Multi-Source Data Store has been developed in 
  the `xcube_multistore` package. It supports the following features:
  * subset of dataset (defined by grid mapping)
  * resample and reproject dataset (defined by grid mapping)
  * grid mapping may be defined by the user or by a dataset 
  * allow for time series at a single spatial point; interpolate the neighbouring points
  * allow data fusion, where data variables in one `xr.Dataset` refers to different data sources
  * support spatial cutout of an area around a defined spatial point.
  * support preload API for [xcube-clms](https://github.com/xcube-dev/xcube-clms) and 
    [xcube-zendoo](https://github.com/xcube-dev/xcube-zenodo)
  * allow to write to netcdf and zarr

xcube_multistore/config.py
# MIT License
#
# Copyright (c) 2025-2026 Brockmann Consult GmbH
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import os
from collections import defaultdict
from typing import Any

import fsspec
import yaml
from xcube.core.store import list_data_store_ids
from xcube.util.jsonschema import (
    JsonArraySchema,
    JsonBooleanSchema,
    JsonComplexSchema,
    JsonNumberSchema,
    JsonObjectSchema,
    JsonStringSchema,
)

from .constants import NAME_WRITE_STORE
from .utils import _remove_compressed_extension

SCHEMA_IDENTIFIER = JsonStringSchema(title="Identifier for the object", min_length=1)

# define schema for dataset
SCHEMA_STORE_IDENTIFIER = JsonStringSchema(
    title="Store identifier used to open the dataset", min_length=1
)
SCHEMA_GRIDMAPPING_ID = JsonStringSchema(
    title="Grid mapping identifier for the dataset", min_length=1
)
SCHEMA_DATA_ID = JsonStringSchema(
    title="Data ID of the dataset in the assigned data store", min_length=1
)
SCHEMA_OPEN_PARAMS = JsonObjectSchema(
    title="Open data parameters", additional_properties=True
)
SCHEMA_XR_MERGE_PARAMS = JsonObjectSchema(
    title="`xarry.merge parameters`",
    description=(
        "See documentation https://docs.xarray.dev/en/stable/generated/xarray.merge.html"
    ),
    default=dict(
        compat="no_conflicts",
        join="exact",
        fill_value="<NA>",
        combine_attrs="drop_conflicts",
    ),
    additional_properties=True,
)
SCHEMA_FORMAT_ID = JsonStringSchema(
    title="Desired format of the saved datacube.",
    default="zarr",
    enum=["netcdf", "zarr", "levels"],
)
SCHEMA_CHUNKSIZE = JsonObjectSchema(
    title="Mapping from dimension to chunk size.",
    required=[],
    additional_properties=True,
)
SCHEMA_CUSTOM_PROCESSING = JsonObjectSchema(
    properties=dict(
        module_path=JsonStringSchema(
            title="Path to python module relative to config file.", min_length=1
        ),
        function_name=JsonStringSchema(
            title="Name of function in the python module.",
            description=(
                "Any function is allowed with takes a `xarray.Dataset` as an "
                "input and returns a modified xarray.Dataset."
            ),
            min_length=1,
        ),
    ),
    required=["module_path", "function_name"],
    additional_properties=False,
)
SCHEMA_SPATIAL_RESAMPLE_PARAMS = JsonObjectSchema(
    title="Resample parameters used for xcube_resampling.resample_in_space",
    description="For further information on the parameters refer to "
    "https://xcube-dev.github.io/xcube-resampling/api/#xcube_resampling.resample_in_space.",
    properties=dict(
        variables=JsonComplexSchema(
            title="A single variable name or iterable of variable names to be "
            "resampled. If None, all data variables will be processed.",
            any_of=[JsonStringSchema(), JsonArraySchema()],
        ),
        interp_methods=JsonComplexSchema(
            title="Optional interpolation method to be used for upsampling spatial "
            "data variables.",
            description="Interpolation method to be used. Can be a single "
            "interpolation method for all variables or a dictionary "
            "mapping variable names or dtypes to interpolation "
            "method.",
            any_of=[
                JsonStringSchema(enum=["nearest", "triangular", "bilinear"]),
                JsonNumberSchema(
                    description="0 => nearest neighbor " "1 => linear/bilinear",
                    enum=[0, 1],
                ),
                JsonObjectSchema(
                    description="A dictionary mapping variable names or dtypes to "
                    "interpolation as shown in the enum in the sibling "
                    "schema.",
                ),
            ],
            default="0 for integer arrays, else 1",
        ),
        agg_methods=JsonComplexSchema(
            title="Optional aggregation methods for downsampling spatial " "variables.",
            any_of=[
                JsonStringSchema(
                    enum=[
                        "center",
                        "count",
                        "first",
                        "last",
                        "max",
                        "mean",
                        "median",
                        "mode",
                        "min",
                        "prod",
                        "std",
                        "sum",
                        "var",
                    ]
                ),
                JsonObjectSchema(
                    description="A dictionary mapping variable names or dtypes to"
                    " methods as shown in the enum in the sibling "
                    "schema."
                ),
            ],
            default="Defaults to 'center' for integer arrays, else 'mean'.",
        ),
        recover_nans=JsonComplexSchema(
            title="Optional boolean or mapping to enable NaN recovery during "
            "upsampling (only applies when interpolation method is not "
            "`nearest`).",
            any_of=[
                JsonBooleanSchema(),
                JsonObjectSchema(
                    description="A dictionary mapping variable names or dtypes to "
                    "booleans"
                ),
            ],
            default="False",
        ),
        fill_values=JsonComplexSchema(
            title="Optional fill value(s) for areas outside input coverage.",
            any_of=[
                JsonNumberSchema(),
                JsonObjectSchema(description="A dictionary by variable or type"),
            ],
            default="Defaults based on data type are used:"
            " `float`: `NaN`, "
            "  `uint8`: `255`, "
            "  `uint16`: `65535`, "
            "  `other ints`: `-1`",
        ),
        tile_size=JsonComplexSchema(
            title="Optional tile size used when generating a regular grid from "
            "an irregular source grid mapping. Only used if `target_gm` is not "
            "provided.",
            any_of=[
                JsonNumberSchema(),
                JsonArraySchema(items=JsonNumberSchema(), min_items=2, max_items=2),
            ],
        ),
    ),
)
SCHEMA_TEMPORAL_RESAMPLE_PARAMS = JsonObjectSchema(
    title="Resample parameters used for xcube_resampling.resample_in_time",
    description="For further information on the parameters refer to "
    "https://xcube-dev.github.io/xcube-resampling/api/#xcube_resampling.resample_in_time.",
    properties=dict(
        frequency=JsonStringSchema(
            title="Target temporal frequency",
            description="Target temporal frequency, following Pandas period aliases. "
            "Format `<count><period>`, where `<period>` may be one of 's', 'min', 'h', "
            "'D', 'W', 'M', 'Q', 'Y'.",
        ),
        variables=JsonComplexSchema(
            title="A single variable name or iterable of variable names to be "
            "resampled. If None, all data variables will be processed.",
            any_of=[JsonStringSchema(), JsonArraySchema()],
        ),
        interp_methods=JsonComplexSchema(
            title="Optional interpolation method(s) for upsampling.",
            any_of=[JsonStringSchema(), JsonArraySchema(), JsonObjectSchema()],
        ),
        agg_methods=JsonComplexSchema(
            title="Optional aggregation method(s) for downsampling.",
            any_of=[JsonStringSchema(), JsonArraySchema(), JsonObjectSchema()],
        ),
        offset=JsonStringSchema(
            title="offset to adjust resampled time labels. Uses the same syntax "
            "as frequency."
        ),
        tolerance=JsonStringSchema(
            title="maximum allowed distance for selective downsampling methods (e.g., "
            "`'backfill'`, `'ffill'`, `'nearest'`). Defaults to the resampling "
            "frequency."
        ),
    ),
)
SCHEMA_DATA_VARIABLE = JsonObjectSchema(
    properties=dict(
        identifier=SCHEMA_IDENTIFIER,
        store=SCHEMA_STORE_IDENTIFIER,
        data_id=SCHEMA_DATA_ID,
        open_params=SCHEMA_OPEN_PARAMS,
        custom_processing=SCHEMA_CUSTOM_PROCESSING,
        spatial_resample_params=SCHEMA_SPATIAL_RESAMPLE_PARAMS,
        temporal_resample_params=SCHEMA_TEMPORAL_RESAMPLE_PARAMS,
    ),
    required=["identifier", "store", "data_id"],
    additional_properties=False,
)
SCHEMA_MULTI_DATASET = JsonObjectSchema(
    properties=dict(
        identifier=SCHEMA_IDENTIFIER,
        variables=JsonArraySchema(
            title="List of data variables",
            items=SCHEMA_DATA_VARIABLE,
        ),
        grid_mapping=SCHEMA_GRIDMAPPING_ID,
        format_id=SCHEMA_FORMAT_ID,
        chunksize=SCHEMA_CHUNKSIZE,
        xr_merge_params=SCHEMA_XR_MERGE_PARAMS,
    ),
    required=["identifier", "variables"],
    additional_properties=False,
)
SCHEMA_SINGLE_DATASET = JsonObjectSchema(
    properties=dict(
        identifier=SCHEMA_IDENTIFIER,
        store=SCHEMA_STORE_IDENTIFIER,
        grid_mapping=SCHEMA_GRIDMAPPING_ID,
        data_id=SCHEMA_DATA_ID,
        open_params=SCHEMA_OPEN_PARAMS,
        format_id=SCHEMA_FORMAT_ID,
        chunksize=SCHEMA_CHUNKSIZE,
        custom_processing=SCHEMA_CUSTOM_PROCESSING,
        spatial_resample_params=SCHEMA_SPATIAL_RESAMPLE_PARAMS,
        temporal_resample_params=SCHEMA_TEMPORAL_RESAMPLE_PARAMS,
    ),
    required=["identifier", "store", "data_id"],
    additional_properties=False,
)
SCHEMA_DATASET = JsonComplexSchema(one_of=[SCHEMA_SINGLE_DATASET, SCHEMA_MULTI_DATASET])

# define schema for data store
SCHEMA_STORE_ID = JsonStringSchema(
    title="Store identifier",
    enum=list_data_store_ids(),
)
SCHEMA_STORE_PARAMS = JsonObjectSchema(
    title="Store parameters",
    description=(
        "Store parameters can be obtained by `get_data_store_params_schema(store_id)`. "
        "Further documentation can be found at "
        "https://xcube.readthedocs.io/en/latest/dataaccess.html#data-store-framework."
    ),
    additional_properties=True,
)
SCHEMA_STORE = JsonObjectSchema(
    properties=dict(
        identifier=SCHEMA_IDENTIFIER,
        store_id=SCHEMA_STORE_ID,
        store_params=SCHEMA_STORE_PARAMS,
    ),
    required=["identifier", "store_id"],
    additional_properties=False,
)

# define schema for grid mapping
SCHEMA_SPATIAL_RES = JsonNumberSchema(title="Spatial Resolution", exclusive_minimum=0.0)
SCHEMA_CRS = JsonStringSchema(title="Coordinate reference system", default="EPSG:4326")
SCHEMA_BBOX = JsonArraySchema(
    items=(
        JsonNumberSchema(),
        JsonNumberSchema(),
        JsonNumberSchema(),
        JsonNumberSchema(),
    ),
    title="Bounding box [west, south, east, north]",
)
SCHEMA_TILE_SIZE = JsonComplexSchema(
    title="Spatial chunk size in grid mapping.",
    one_of=[
        JsonArraySchema(
            items=(
                JsonNumberSchema(title="Chunk size in lat/x direction."),
                JsonNumberSchema(title="Chunk size in lon/y direction."),
            ),
        ),
        JsonNumberSchema(title="Squared chunk size"),
    ],
    default=1024,
)

SCHEMA_GRID_MAPPING = JsonObjectSchema(
    properties=dict(
        identifier=SCHEMA_IDENTIFIER,
        bbox=SCHEMA_BBOX,
        crs=SCHEMA_CRS,
        spatial_res=SCHEMA_SPATIAL_RES,
        tile_size=SCHEMA_TILE_SIZE,
    ),
    required=["identifier", "bbox", "crs", "spatial_res"],
    additional_properties=False,
)

# define schema for preload dataset
SCHEMA_PRELOAD_PARAMS = JsonObjectSchema(
    title="Preload data parameters", additional_properties=True
)
SCHEMA_PRELOAD_DATASET = JsonObjectSchema(
    properties=dict(
        store=SCHEMA_STORE_IDENTIFIER,
        data_ids=JsonArraySchema(items=SCHEMA_DATA_ID),
        preload_params=SCHEMA_PRELOAD_PARAMS,
    ),
    required=["store", "data_ids"],
    additional_properties=False,
)


# define schema for general
SCHEMA_VISUALIZE = JsonBooleanSchema(
    title="Switch between visualization in table, if True, and logging, if False.",
    default=True,
)
SCHEMA_FORCE_PRELOAD = JsonBooleanSchema(
    description="If True, all data IDS given in section `preload_datasets` will be "
    "preloaded. If False, only non-preloaded datasets will be preloaded.",
    default=False,
)
SCHEMA_GDAL_HTTP_PARAMS = JsonObjectSchema(
    properties=dict(
        gdal_http_max_retry=JsonNumberSchema(
            title="Maximal number of retries of a http request in GDAL.",
            description=(
                "This is used when opening a tif file with `rioxarray.open_rasterio` "
                "from a remote source."
            ),
            default=10,
        ),
        gdal_http_retry_delay=JsonNumberSchema(
            title="Delay in seconds between retries of a http request in GDAL.",
            description=(
                "This is used when opening a tif file with `rioxarray.open_rasterio` "
                "from a remote source."
            ),
            default=5,
        ),
    ),
    required=[],
    additional_properties=False,
)
SCHEMA_GENERAL = JsonObjectSchema(
    properties=dict(
        visualize=SCHEMA_VISUALIZE,
        force_preload=SCHEMA_FORCE_PRELOAD,
        gdal_http_params=SCHEMA_GDAL_HTTP_PARAMS,
    ),
    required=[],
    additional_properties=False,
)

# define general config schema
CONFIG_SCHEMA = JsonObjectSchema(
    properties=dict(
        datasets=JsonArraySchema(items=SCHEMA_DATASET),
        preload_datasets=JsonArraySchema(items=SCHEMA_PRELOAD_DATASET),
        data_stores=JsonArraySchema(
            items=SCHEMA_STORE,
            title="Used xcube's data stores.",
            description=(
                f"A single data store with identifier {NAME_WRITE_STORE!r} is "
                f"required for writing the final data cubes."
            ),
        ),
        grid_mappings=JsonArraySchema(items=SCHEMA_GRID_MAPPING),
        general=SCHEMA_GENERAL,
    ),
    required=["datasets", "data_stores"],
    additional_properties=False,
)


class MultiSourceConfig:

    def __init__(self, config: str | dict[str, Any]):
        if not isinstance(config, dict):
            config = _read_yaml(config)
        schema = self.get_schema()
        schema.validate_instance(config)
        self.original: dict = config
        self.preload_datasets = config.get("preload_datasets", None)
        self.datasets = {
            config_ds["identifier"]: config_ds for config_ds in config["datasets"]
        }
        self.preload_map = self._get_preload_map()
        self.data_stores = {
            config_store["identifier"]: config_store
            for config_store in config["data_stores"]
        }
        grid_mappings = config.get("grid_mappings")
        if grid_mappings:
            self.grid_mappings = {
                config_gm["identifier"]: config_gm for config_gm in grid_mappings
            }
        else:
            self.grid_mappings = None
        self.general = config.get("general", {})
        self._general_setup()

        assert (
            NAME_WRITE_STORE in self.data_stores.keys()
        ), f"store with identifier {NAME_WRITE_STORE!r} needs to be given."

    def _general_setup(self):
        if "visualize" not in self.general:
            self.general["visualize"] = True
        if "force_preload" not in self.general:
            self.general["force_preload"] = False
        _GDAL_HTTP_MAX_RETRY_DEFAULT = 10
        _GDAL_HTTP_RETRY_DELAY_DEFAULT = 5
        if "gdal_http_params" not in self.general:
            self.general["gdal_http_params"] = dict(
                gdal_http_max_retry=_GDAL_HTTP_MAX_RETRY_DEFAULT,
                gdal_http_retry_delay=_GDAL_HTTP_RETRY_DELAY_DEFAULT,
            )
        if "gdal_http_max_retry" not in self.general["gdal_http_params"]:
            self.general["gdal_http_params"][
                "gdal_http_max_retry"
            ] = _GDAL_HTTP_MAX_RETRY_DEFAULT
        if "gdal_http_retry_delay" not in self.general["gdal_http_params"]:
            self.general["gdal_http_params"][
                "gdal_http_retry_delay"
            ] = _GDAL_HTTP_RETRY_DELAY_DEFAULT

        # apply general setup
        os.environ["GDAL_HTTP_MAX_RETRY"] = str(
            self.general["gdal_http_params"]["gdal_http_max_retry"]
        )
        os.environ["GDAL_HTTP_RETRY_DELAY"] = str(
            self.general["gdal_http_params"]["gdal_http_retry_delay"]
        )

    def _get_preload_map(self):
        if self.preload_datasets is None:
            return None
        preload_map = defaultdict(list)
        for config_preload in self.preload_datasets:
            for data_id in config_preload["data_ids"]:
                data_id_mod = _remove_compressed_extension(data_id)
                for identifier_ds, config_ds in self.datasets.items():
                    if "variables" in config_ds:
                        for config_da in config_ds["variables"]:
                            if data_id_mod in config_da["data_id"]:
                                preload_map[data_id].append(config_da["data_id"])
                    else:
                        if data_id_mod in config_ds["data_id"]:
                            preload_map[data_id].append(config_ds["data_id"])
        return preload_map

    @classmethod
    def get_schema(cls) -> JsonObjectSchema:
        return CONFIG_SCHEMA


def _read_yaml(config_path: str) -> dict[str, Any]:
    with fsspec.open(config_path, "r") as file:
        config = yaml.safe_load(file)
    return config


def _is_jupyter_notebook():
    try:
        from IPython import get_ipython

        return get_ipython() is not None
    except ModuleNotFoundError:
        return False
